### PR TITLE
Moved Cohere import

### PR DESCRIPTION
**Pull request by ivanleomk**
| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 0f949bdec9371e467f28998a95289235acf917c9  | 
|--------|--------|

### Summary:
Moved `NonStreamedChatResponse` import, updated dependencies, fixed import paths, enhanced retry functions, and added tests.

**Key points**:
- Moved `NonStreamedChatResponse` import inside `parse_cohere_json_schema` in `instructor/function_calls.py`.
- Updated `pyproject.toml` to include `cohere` as an optional dependency.
- Fixed import path in `tests/llm/test_openai/schema/test_async_decorator.py` from `.util` to `..util`.
- Updated `retry_sync` and `retry_async` in `instructor/retry.py` to handle `chat_history` and `contents` in `kwargs`.
- Added `tests/llm/test_cohere/test_retries.py` for testing Cohere retry logic.
- Unmarked several tests in `tests/llm/test_openai/test_retries.py` by removing `@pytest.mark.skip()`.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by ivanleomk** [comment]
This removes the top level cohere import from the `function_calls` file, allowing users to use instructor without having to install cohere

File: instructor/function_calls.py [hunk 1]
```diff
@@ -276,10 +276,13 @@ def from_response(
...
     def parse_cohere_json_schema(
         cls: type[BaseModel],
-        completion: NonStreamedChatResponse,
+        completion: Any,
         validation_context: Optional[dict[str, Any]] = None,
         strict: Optional[bool] = None,
     ):
+        from cohere import NonStreamedChatResponse
+
+        assert isinstance(completion, NonStreamedChatResponse)
```
Comment (jxnl) [inline instructor/function_calls.py:285]:
```suggestion
        assert isinstance(completion, NonStreamedChatResponse), "completion is not of type `NonStreamedChatResponse`"
```

Comment (jxnl) [reply instructor/function_calls.py:285]: but maybe better to just check if hasattr("text")

Comment (ivanleomk) [reply instructor/function_calls.py]: converted to 

```
 assert hasattr(
            completion, "text"
        ), "Completion is not of type NonStreamedChatResponse"
```

**Review by jxnl — APPROVED** [review]

**Pull request merged**